charms/knative-serving/src/charm_template.py

```diff
@@ -0,0 +1,248 @@
+        pass
+
+    def on_upgrade_charm(self, event):
+        self.on_install(event)
+
+    def on_update_status(self, event):
+        charm_status = self.charm_status()
+        if not isinstance(charm_status, ActiveStatus):
+            # Queue an install event to try to reconcile our resources
+            self.log.info("Charm emitting an install event to try to reconcile resources")
+        self.on.install.emit()
+        self.log_and_set_status(charm_status)
+
+    def reconcile(self, resources: Optional[List[Union[NamespacedResource, GlobalResource]]] = None):
+        """Reconcile our Kubernetes objects to achieve the desired state
+
+        This can be invoked to both install or update objects in the cluster.  It uses an apply
+        logic to update things only if necessary.  This method by default __does not__ remove
+        objects that are no longer required - if handling that situation is required, it must be
+        done separately.
+        """
+        self.log.info("Reconciling")
+        if resources is None:
+            resources = self.render_manifests()
+        self.log.debug(f"Applying {len(resources)} resources")
+
+        try:
+            self.lightkube_client.apply_many(resources)
+        except ApiError as e:
+            # Handle fobidden error as this likely means we do not have --trust
+            if e.status.code == 403:
+                self.logger.error(
+                    "Received Forbidden (403) error from lightkube when creating resources.  "
+                    "This may be due to the charm lacking permissions to create cluster-scoped "
+                    "roles and resources.  Charm must be deployed with `--trust`"
+                )
+                self.logger.error(f"Error received: {str(e)}")
+                raise ReconcileError("Cannot create required resources.  Charm may be missing `--trust`", BlockedStatus)
+            else:
+                raise e
+        self.log.info("Reconcile completed successfully")
+
+    def render_manifests(self) -> List[Union[NamespacedResource, GlobalResource]]:
+        """Renders this charm's manifests, returning them as a list of Lightkube Resources
+
+        If overriding this class, you should replace it with a method that will always generate
+        a list of all resources that should currently exist in the cluster.
+        """
+        self.log.info("Rendering manifests")
+        self.log.debug(f"Rendering with context: {self.context_for_render}")
+        manifest_parts = []
+        for template_file in self.template_files:
+            self.log.debug(f"Rendering manifest for {template_file}")
+            manifest_parts.append(self.jinja_env.get_template(template_file).render(**self.context_for_render))
+            self.log.debug(f"Rendered manifest:\n{manifest_parts[-1]}")
+        return codecs.load_all_yaml("\n---\n".join(manifest_parts))
```

Comment:
How does this behave with preexisting `---` in the manifests? I'm worried that this approach might be a bit fragile.